plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java

```diff
@@ -0,0 +1,272 @@
+
+            getFutureValue(allOf(
+                    writeChunkFutures.toArray(CompletableFuture[]::new)
+            ));
+
+            final Map<String, HiveColumnStatistics> currentTableColumnStatistics = this.getTableColumnStatistics(table);
+            Set<String> removedStatistics = difference(currentTableColumnStatistics.keySet(), updatedTableColumnStatistics.keySet());
+            List<CompletableFuture<Void>> writeFutures = removedStatistics.stream()
+                    .map(column -> runAsync(() ->
+                            glueClient.deleteColumnStatisticsForTable(
+                                    new DeleteColumnStatisticsForTableRequest()
+                                            .withCatalogId(catalogId)
+                                            .withDatabaseName(table.getDatabaseName())
+                                            .withTableName(table.getTableName())
+                                            .withColumnName(column)))
+                    ).collect(toUnmodifiableList());
+
+            for (CompletableFuture<Void> writeFuture : writeFutures) {
+                getFutureValue(writeFuture);
+            }
+        }
+        catch (AmazonServiceException ex) {
+            throw new TrinoException(HIVE_METASTORE_ERROR, ex);
+        }
+    }
+
+    @Override
+    public void updatePartitionStatistics(Partition partition, Map<String, HiveColumnStatistics> updatedColumnStatistics)
+    {
+        try {
+            HiveBasicStatistics partitionStats = getHiveBasicStatistics(partition.getParameters());
+            List<ColumnStatistics> columnStats = toGlueColumnStatistics(partition, updatedColumnStatistics, partitionStats.getRowCount());
+
+            List<List<ColumnStatistics>> columnChunks = Lists.partition(columnStats, GLUE_COLUMN_WRITE_STAT_PAGE_SIZE);
+
+            List<CompletableFuture<Void>> writePartitionStatsFutures = columnChunks.stream()
+                    .map(columnChunk ->
+                            runAsync(() -> glueClient.updateColumnStatisticsForPartition(new UpdateColumnStatisticsForPartitionRequest()
+                                    .withCatalogId(catalogId)
+                                    .withDatabaseName(partition.getDatabaseName())
+                                    .withTableName(partition.getTableName())
+                                    .withPartitionValues(partition.getValues())
+                                    .withColumnStatisticsList(columnChunk)), writeExecutor)
+                    ).collect(toUnmodifiableList());
+
+            for (CompletableFuture<Void> writePartitionStatsFuture : writePartitionStatsFutures) {
+                getFutureValue(writePartitionStatsFuture);
+            }
+
+            boolean partitionExists = partitionExists(partition);
+            Map<String, HiveColumnStatistics> currentColumnStatistics = partitionExists ? this.getPartitionColumnStatistics(partition) : Collections.emptyMap();
```

Comment:
What happens if `getPartitionStatistics` is called for a non-existent partition?

There will be 1 API call on the not exists path but 2 on the exists path. Can we avoid this by using the exception thrown when a partition doesn't exist as a control variable instead of the boolean?

Lots of people run into Glue rate-limits so we should strive to reduce API calls to Glue.

Reply:
When called on a non existing partition glue client throw an exception, we may use the exception to check if the partition is present as it will reduce the number of api calls.
In `getPartitionColumnStatistics` we should check for `EntityNotFoundException` raised by glue client and in that case return an empty partition statistics object, do you think if may be a good compromise between readability and api calls count ?